tools/wake-format/emitter.cpp

```diff
@@ -279,6 +270,147 @@ static wcl::doc binop_rhs_separator(const CSTElement& op) {
+  collect.push_back(right);
+
+  return collect;
+}
+
+static std::vector<CSTElement> collect_right_binary(CSTElement collect_over, CSTElement node) {
+  if (node.id() != CST_BINARY) {
+    return {node};
+  }
+
+  // NOTE: The 'node' variant functions are being used here which is differnt than everywhere else
+  // This is fine since COMMENTS are bound to the nodes and this func only needs to process nodes
+  CSTElement left = node.firstChildNode();
+  CSTElement op = left;
+  op.nextSiblingNode();
+  CSTElement right = op;
+  right.nextSiblingNode();
+
+  if (!(op.id() == CST_OP && op.firstChildElement().id() == collect_over.id() &&
+        op.firstChildElement().fragment().segment().str() ==
+            collect_over.fragment().segment().str())) {
+    return {node};
+  }
+
+  std::vector<CSTElement> collect = {left};
+  auto right_collect = collect_right_binary(collect_over, right);
+
+  collect.insert(collect.end(), right_collect.begin(), right_collect.end());
+
+  return collect;
+}
+
+static std::vector<CSTElement> collect_apply_parts(CSTElement node) {
+  if (node.id() != CST_APP) {
+    return {node};
+  }
+
+  // NOTE: The 'node' variant functions are being used here which is differnt than everywhere else
+  // This is fine since COMMENTS are bound to the nodes and this func only needs to process nodes
+  CSTElement lhs = node.firstChildNode();
+  CSTElement rhs = lhs;
+  rhs.nextSiblingNode();
+
+  auto collect = collect_apply_parts(lhs);
+  collect.push_back(rhs);
+
+  return collect;
+}
+
+static inline bool is_simple_binop(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node,
+                                   const token_traits_map_t& traits) {
+  if (node.id() != CST_BINARY) {
+    return is_simple_literal(builder, ctx, node, traits);
+  }
+
+  // NOTE: The 'node' variant functions are being used here which is differnt than everywhere else
+  // This is fine since COMMENTS are bound to the nodes and this func only needs to process nodes
+  CSTElement lhs = node.firstChildNode();
+  CSTElement op = lhs;
+  op.nextSiblingNode();
+  CSTElement rhs = op;
+  rhs.nextSiblingNode();
+
+  FMT_ASSERT(op.id() == CST_OP, op, "Expected CST_OP for operator");
+  CSTElement op_token = op.firstChildElement();
+
+  std::vector<CSTElement> parts = {};
+  if (is_op_left_assoc(op_token)) {
+    parts = collect_left_binary(op_token, node);
+  } else {
+    parts = collect_right_binary(op_token, node);
+  }
+
+  if (parts.size() != 2) {
+    return false;
+  }
+
+  return is_simple_literal(builder, ctx, parts[0], traits) &&
+         is_simple_literal(builder, ctx, parts[1], traits);
+}
+
+static inline bool is_simple_apply(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node,
+                                   const token_traits_map_t& traits) {
+  if (node.id() != CST_APP) {
+    return is_simple_literal(builder, ctx, node, traits);
+  }
+
+  auto parts = collect_apply_parts(node);
+
+  if (parts.size() != 2) {
```

Comment:
I don't expect there to ever be a `CST_APP` with less than two parts, but maybe `>` for better semantic readability anyway?
```suggestion
  if (parts.size() > 2) {
```